### ISSUE TITLE
rfc(git-clone) update repo and post pull script

### ISSUE THREAD

**Issue by djarbz** [coder/modules#332]
Looking for thoughts on these two items, I'm willing to submit a PR if we like it.

1. If the repo already exists, we should fetch updates and perform a ff merge on an opt-in basis.
Something like below, I just added this to my dotfiles.
Might need to rework the merge logic to select the correct branch.
```bash
if [ -v PROJECT_DIRECTORY ]; then
  echo "Checking for project updates"
  current_branch=$(git rev-parse --abbrev-ref HEAD)
  # Sync remote changes
  git -C "${PROJECT_DIRECTORY}" fetch origin
  # Apply updates if no conflicts
  git -C "${PROJECT_DIRECTORY}" merge "origin/${current_branch}" --ff-only
fi
```

2. After successfully cloning, or performing a ff merge, we should run a "setup script" if found to perform repo specific setup.
This is currently what I have in my dotfiles, doesn't matter if the script is executable or not.
```bash
# Prep cloned workspace via any found `coder.bootstrap` executable files in the project directory.
if [ -v PROJECT_DIRECTORY ]; then
  echo "Prepping Project Workspace..."
  find "${PROJECT_DIRECTORY}" -name coder.bootstrap -type f -print -exec bash {} \;
fi
```

**Comment by matifali** [comment]
I agree with the proposal if we also add a switch to retain the previous behaviour of the git-clone module. 
@johnstcn what do you think?

**Comment by johnstcn** [comment]
Re: 1) I think this is fine on an opt-in basis

Re: 2) This could be done as a separate step; there may be cases where this requires specific dependencies that are not yet available inside the workspace?

**Comment by matifali** [comment]
For me, 2) is out of scope for git-clone.

**Comment by gjrtimmer** [comment]
I like this, working on the same thing for myself; I would like to offer a suggestion: instead of all the different file locations that everybody is using, why not a standardized approach, I prefer myself the location of `.coder/bootstrap` or `.coder/bootstrap.sh` in my repositories, gives more flexibility for the future and allows for multiple files/scripts regarding coder workspaces to be organized in a `.coder` folder just like other platforms have either own folders, `.gitlab` `.github` etc. What do yo think?

**Comment by djarbz** [comment]
Sorry I've been MIA for a month.

RE 2) I think it would still be in scope, at least until we can figure out module dependencies. If we added it as a separate module, then we wouldn't be able to tell if the clone completed or not.

I think having a `.coder` directory makes sense, then we could also match based on the other "init" style scripts that coder uses elsewhere.
